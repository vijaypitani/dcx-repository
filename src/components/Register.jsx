import { Link } from "react-router-dom";
import React, { Component } from "react";
import DeveloperService from "../services/DeveloperService";
import Header from "./Header";

const INITIAL_STATE = {
  full_name: "",
  password: "",
  email: "",
  group: "",
  full_nameError: "",
  passwordError: "",
  emailError: "",
  groupError: "",
  message1: "",
  message2: "",
};

class Register extends Component {
  constructor(props) {
    super(props);

    this.state = INITIAL_STATE;

    this.stopSubmission = this.stopSubmission.bind(this);
  }

  validate = () => {
    let full_nameError = "";
    let passwordError = "";
    let emailError = "";
    let groupError = "";

    if (!this.state.full_name) {
      full_nameError = "Name cannot be empty";
    } else if (!this.state.full_name.match(/^[a-zA-Z\s]+$/)) {
      full_nameError = "Name must include only alphabets";
    }
    if (!this.state.password) {
      passwordError = "Password cannot be empty";
    } else if (!this.state.password.match("(?=.*[@#$]).{8,}")) {
      passwordError =
        "Password must be at least 8 characters with at least one alphabet,digit and two special characters";
    }
    if (!this.state.email) {
      emailError = "Email cannot be empty";
    } else if (
      !this.state.email.match(/[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$/)
    ) {
      emailError = "Email is invalid";
    }
    if (!this.state.group) {
      groupError = "Group cannot be empty";
    }
    if (full_nameError || passwordError || emailError || groupError) {
      this.setState({
        full_nameError,
        passwordError,
        emailError,
        groupError,
      });
      return false;
    }
    return true;
  };

  stopSubmission = (e) => {
    e.preventDefault();
    const data = {
      full_name: this.state.full_name,
      email: this.state.email,
      password: this.state.password,
      group: this.state.group,
    };

    const isValid = this.validate();
    if (isValid) {
      DeveloperService.registerUser(data)
        .then((res) => {
          this.setState({
            message1: "Developer Added Successfully",
          });
        })
        .catch((err) => {
          this.setState({
            message2: err.response.data.message,
          });
        });
      // alert('Developer added Successfully')
      console.log("Developer Details:", this.state);
      this.setState(INITIAL_STATE);
    }
  };

  render() {
    let error = "";
    let status = "";
    if (this.state.message2) {
      error = (
        <div className="alert alert-danger text-center" role="alert">
          {this.state.message2}
        </div>
      );
    }
    if (this.state.message1) {
      status = (
        <div className="alert alert-success text-center" role="alert">
          {this.state.message1}
        </div>
      );
    }

    return (
      <div>
        <div id="container">
          <Header />
        </div>
        <div id="container">
          <nav id="leftMenu">
            <h3>Links</h3>
            <ul>
              <li>
                <Link to="/dashboard">SEO</Link>
              </li>
              <li>
                <Link to="/dashboard">PHP</Link>
              </li>
              <li>
                <Link to="/dashboard">Ajax</Link>
              </li>
              <li>
                <Link to="/dashboard">jQuery</Link>
              </li>
              <li>
                <Link to="/dashboard">Web design</Link>
              </li>
              <li>
                <Link to="/dashboard">Web Programming</Link>
              </li>
              <li>
                <Link to="/dashboard">Content Creation</Link>
              </li>
              <li>
                <Link to="/dashboard">Internet Marketing</Link>
              </li>
              <li>
                <Link to="/dashboard">XHTML Templates</Link>
              </li>
            </ul>
          </nav>

          <section>
            <h2>Register A Developer</h2>
            <div>Please use this form to add a new developer</div>
            <br></br>

            <form
              method="POST"
              onSubmit={this.stopSubmission}
              data-testid="form"
            >
              {error}
              {status}
              <div>
                <label>Full Name:</label>
                <input
                  name="full_name"
                  className="form-control"
                  placeholder="eg. John Doe"
                  value={this.state.full_name}
                  onChange={(e) => this.setState({ full_name: e.target.value })}
                />
              </div>

              <div className="text-danger">{this.state.full_nameError}</div>
              <br></br>
              <div>
                <label>Email:</label>
                <input
                  name="email"
                  className="form-control"
                  placeholder="eg. john.doe@gmail.com"
                  value={this.state.email}
                  onChange={(e) => this.setState({ email: e.target.value })}
                />
              </div>
              <div className="text-danger">{this.state.emailError}</div>
              <br></br>
              <div>
                <label>Password</label>
                <input
                  name="password"
                  type="password"
                  className="form-control"
                  placeholder="John*123#"
                  value={this.state.password}
                  onChange={(e) => this.setState({ password: e.target.value })}
                />
              </div>
              <div className="text-danger">{this.state.passwordError}</div>
              <br></br>
              <div>
                <label>Group</label>
              </div>

              <select
                className="custom-select"
                name="group"
                value={this.state.group}
                onChange={(e) => this.setState({ group: e.target.value })}
              >
                <option value="" disabled={true} defaultValue={true}>
                  Select group
                </option>
                <option>Admin</option>
                <option>Developer</option>
              </select>

              <div className="text-danger">{this.state.groupError}</div>
              <br></br>
              <div className="text-center">
                <button
                  className="btn btn-outline-success mr-md-4 btn-sm"
                  disabled={!this.state.group}
                >
                  Submit
                </button>
                <Link to="/dashboard">
                  <button className="btn btn-outline-secondary btn-sm">
                    Go Back
                  </button>
                </Link>
              </div>
            </form>
          </section>
        </div>
        <div style={{ clear: "both" }}></div>
        <br></br>
        <br></br>
        <footer>Copyright 2017. DCX Developer Directory.</footer>
      </div>
    );
  }
}

export default Register;
