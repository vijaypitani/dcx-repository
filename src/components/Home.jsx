import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { Redirect } from "react-router";
import About from "./About";
import Dashboard from "./Dashboard";
import Register from "./Register";
import Browse from "./Browse";
import Contact from "./Contact";
import Login from "./Login";
import SignUp from "./SignUp";
import ForgotPassword from "./ForgotPassword";
import NewPassword from "./NewPassword";
import React, { useState, useCallback, useEffect } from "react";
import axios from "axios";
import Not_Found from "./Not_Found";

export const UserContext = React.createContext();

function Home() {
  const [role, setRole] = useState("");
  const [auth, setAuth] = useState(false);

  const LoginGuardRoute = ({ component: Component, ...props }) => (
    <Route
      {...props}
      render={(routeProps) => {
        return localStorage.getItem("login") ? (
          <Component {...routeProps} />
        ) : (
          <Redirect to="/login" />
        );
      }}
    />
  );

  const roleAuth = useCallback(async () => {
    let loginInstance = JSON.parse(localStorage.getItem("login"));
    if (loginInstance) {
      //console.log(loginInstance.token);
      if (loginInstance.token) {
        await axios
          .get(
            `https://dcx-server.herokuapp.com/api/verify/data?token=${loginInstance.token}`
          )
          .then((res) => {
            if (res) {
              // console.log("MESSAGE", res.data.message);
              // console.log("DATA", res.data.data);
              // console.log("GROUP", res.data.data.group);
              setRole(res.data.data.group);
              setAuth(true);
            }
          })
          .catch((err) => {
            //console.log(err.response.data.message);
            setRole("");
            setAuth(false);
          });
      }
    }
  }, []);

  useEffect(() => {
    roleAuth();
  }, [roleAuth]);

  // if (role === "" && auth === false) {
  //   return <div>Verifying...</div>;
  // }

  return (
    <UserContext.Provider value={{ Role: role, Auth: auth }}>
      <Router>
        <div>
          <Switch>
            <LoginGuardRoute exact path="/" component={Dashboard} />
            <LoginGuardRoute exact path="/dashboard" component={Dashboard} />
            <LoginGuardRoute exact path="/browse" component={Browse} />
            <Route exact path="/login" component={Login}></Route>
            <Route exact path="/signup" component={SignUp}></Route>
            <Route exact path="/reset" component={ForgotPassword}></Route>
            <Route exact path="/reset/:token" component={NewPassword}></Route>
            <LoginGuardRoute exact path="/about" component={About} />
            <LoginGuardRoute path="/register" component={Register} />
            <LoginGuardRoute path="/contact" component={Contact} />
            <Route component={Not_Found} />
          </Switch>
        </div>
      </Router>
    </UserContext.Provider>
  );
}

export default Home;
