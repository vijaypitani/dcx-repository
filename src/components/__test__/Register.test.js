import { render, screen } from '@testing-library/react';
import ReactDOM from 'react-dom'
import { BrowserRouter as Router } from 'react-router-dom';
import Register from '../Register';
import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount } from 'enzyme';
import { shallow, configure } from 'enzyme';
configure({ adapter: new Adapter() });
import { UserContext } from '../Home'


describe("Register", () => {
  let wrapper;
  let stopSubmission;
  beforeEach(() => {
    stopSubmission = jest.fn();
    wrapper = shallow(<Register submit={stopSubmission} />);
  });

  //Snapshot Test
  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  //Test if a button is disabled
  it('renders register button', () => {
    const div = document.createElement("div");
    ReactDOM.render(<Router>    <UserContext.Provider value={{ Role: "Admin", Auth: false }}>
      <Register />
    </UserContext.Provider></Router>, div)
    const button = div.querySelector('button');
    expect(button).toBeDisabled();
  });



  it('calls stopSubmission function when form is submitted', () => {
    const stopSubmission = jest.fn();
    const wrapper = mount(<Router>    <UserContext.Provider value={{ Role: "Admin", Auth: false }}>
      <Register onSubmit={stopSubmission} />
    </UserContext.Provider></Router>);
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(stopSubmission).toHaveBeenCalledTimes(0);
  });

  it('renders text input with label (default type)', () => {
    const wrapper = mount(<Router> <UserContext.Provider value={{ Role: "Admin", Auth: false }}><Register name="full_name" label="Full Name:" /></UserContext.Provider></Router>);
    const label = wrapper.find('label');
    expect(label).toHaveLength(4);
    const input = wrapper.find('input');
    expect(input).toHaveLength(3);

  });

  it('renders buttons', () => {
    const wrapper = mount(<Router><UserContext.Provider value={{ Role: "Admin", Auth: false }}><Register >Submit</Register></UserContext.Provider></Router>);
    const button = wrapper.find('button');
    expect(button).toHaveLength(2);

  });
  it('renders select with custom text', () => {
    const wrapper = mount(<Router><UserContext.Provider value={{ Role: "Admin", Auth: false }}><Register >Submit</Register></UserContext.Provider></Router>);
    const select = wrapper.find('select');
    expect(select).toHaveLength(1);
    expect(select.prop('name')).toEqual('group');
    expect(select.prop('className')).toEqual('custom-select');
    expect(select.text()).toEqual('Select groupAdminDeveloper');

  });
})

