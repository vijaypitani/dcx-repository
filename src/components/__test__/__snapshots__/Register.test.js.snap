// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should match the snapshot 1`] = `ShallowWrapper {}`;
